spec/std/path_spec.cr

```diff
@@ -0,0 +1,558 @@
+        it_normalizes_path("abc/def/..", "abc", "abc")
+        it_normalizes_path("abc/def/../..", ".", ".")
+        it_normalizes_path("/abc/def/../..", "/", "\\")
+        it_normalizes_path("abc/def/../../..", "..", "..")
+        it_normalizes_path("/abc/def/../../..", "/", "\\")
+        it_normalizes_path("abc/def/../../../ghi/jkl/../../../mno", "../../mno", "..\\..\\mno")
+      end
+
+      describe "combinations" do
+        it_normalizes_path("abc/./../def", "def", "def")
+        it_normalizes_path("abc//./../def", "def", "def")
+        it_normalizes_path("abc/../../././../def", "../../def", "..\\..\\def")
+      end
+    end
+
+    describe "paths with backslash" do
+      describe "already clean" do
+        it_normalizes_path("abc\\def")
+        it_normalizes_path("a\\b\\c")
+        it_normalizes_path("..\\..")
+        it_normalizes_path("..\\..\\abc")
+        it_normalizes_path("\\abc")
+        it_normalizes_path("\\")
+      end
+
+      describe "removes trailing slash" do
+        it_normalizes_path("abc\\", windows: "abc")
+        it_normalizes_path("abc\\def\\", windows: "abc\\def")
+        it_normalizes_path("a\\b\\c\\", windows: "a\\b\\c")
+        it_normalizes_path(".\\", windows: ".")
+        it_normalizes_path("..\\", windows: "..")
+        it_normalizes_path("..\\..\\", windows: "..\\..")
+        it_normalizes_path("\\abc\\", windows: "\\abc")
+      end
+
+      describe "removes double slash" do
+        it_normalizes_path("abc\\\\def\\\\ghi", windows: "abc\\def\\ghi")
+        it_normalizes_path("\\\\abc", windows: "\\abc")
+        it_normalizes_path("\\\\\\abc", windows: "\\abc")
+        it_normalizes_path("\\\\abc\\\\", windows: "\\abc")
+        it_normalizes_path("abc\\\\", windows: "abc")
+      end
+
+      describe "removes ." do
+        it_normalizes_path("abc\\.\\def", windows: "abc\\def")
+        it_normalizes_path("\\.\\abc\\def", windows: "\\abc\\def")
+        it_normalizes_path("abc\\.", windows: "abc")
+      end
+
+      describe "removes .." do
+        it_normalizes_path("abc\\def\\ghi\\..\\jkl", windows: "abc\\def\\jkl")
+        it_normalizes_path("abc\\def\\..\\ghi\\..\\jkl", windows: "abc\\jkl")
+        it_normalizes_path("abc\\def\\..", windows: "abc")
+        it_normalizes_path("abc\\def\\..\\..", windows: ".")
+        it_normalizes_path("\\abc\\def\\..\\..", windows: "\\")
+        it_normalizes_path("abc\\def\\..\\..\\..", windows: "..")
+        it_normalizes_path("\\abc\\def\\..\\..\\..", windows: "\\")
+        it_normalizes_path("abc\\def\\..\\..\\..\\ghi\\jkl\\..\\..\\..\\mno", windows: "..\\..\\mno")
+      end
+
+      describe "combinations" do
+        it_normalizes_path("abc\\.\\..\\def", windows: "def")
+        it_normalizes_path("abc\\\\.\\..\\def", windows: "def")
+        it_normalizes_path("abc\\..\\..\\.\\.\\..\\def", windows: "..\\..\\def")
+      end
+    end
+  end
+
+  describe "#join" do
```

Comment:
what happen if join is performed over absolute paths arguments other than the first one?
what happen if join is performed over paths of different kind?

Reply:
First question is answered in [line 424](https://github.com/straight-shoota/crystal/blob/d9a438a902ae0e9701a529161bd34b3467bffe49/spec/std/path_spec.cr#L424).

Second one: It behaves as if it was implicitly converted to the same kind. I've added specs for this and a missing conversion from windows to posix separators.

Reply:
I think there is [one conversion](https://github.com/crystal-lang/crystal/pull/5635#pullrequestreview-134322174) still missing. But other than that LGTM!